korangar/src/loaders/animation/mod.rs

```diff
@@ -0,0 +1,474 @@
+            frame.upleft = frame.offset - half_size;
+        }
+        // If there is no frame return an image with 1 pixel.
+        if frames.is_empty() {
+            let frame_part = FramePart {
+                animation_index: -1,
+                sprite_number: -1,
+                size: Vector2::new(1, 1),
+                upleft: Vector2::zero(),
+                offset: Vector2::zero(),
+                mirror: false,
+                angle: 0.0,
+                color: Color {
+                    red: 0.0,
+                    blue: 0.0,
+                    green: 0.0,
+                    alpha: 0.0,
+                },
+            };
+            let frame = Frame {
+                size: Vector2::new(1, 1),
+                upleft: Vector2::zero(),
+                offset: Vector2::zero(),
+                remove_offset: Vector2::zero(),
+                frameparts: vec![frame_part],
+            };
+            return frame;
+        }
+        // Find the upmost and leftmost coordinates
+        let upleft_x = frames.iter().min_by_key(|frame| frame.upleft.x).unwrap().upleft.x;
+        let upleft_y = frames.iter().min_by_key(|frame| frame.upleft.y).unwrap().upleft.y;
+
+        // Find the downmost and rightmost coordinates
+        let frame_x = frames.iter().max_by_key(|frame| frame.upleft.x + frame.size.x as i32).unwrap();
+        let frame_y = frames.iter().max_by_key(|frame| frame.upleft.y + frame.size.y as i32).unwrap();
+
+        // Calculate the new rectangle that is formed
+        let mut new_width = frame_x.upleft.x + frame_x.size.x as i32;
+        let mut new_height = frame_y.upleft.y + frame_y.size.y as i32;
+        new_width -= upleft_x;
+        new_height -= upleft_y;
+
+        let mut new_frameparts = Vec::<FramePart>::new();
+        for index in 0..frames.len() {
+            new_frameparts.append(&mut frames[index].frameparts);
+        }
+
+        /* As the origin is (0,0), you get the upleft point of the retangle and
+         * shift to the center, the offset is the difference between the origin and
+         * this point */
+        Frame {
+            size: Vector2::new(new_width, new_height),
+            upleft: Vector2::zero(),
+            offset: Vector2::new(upleft_x + (new_width - 1) / 2, upleft_y + (new_height - 1) / 2),
+            remove_offset: Vector2::zero(),
+            frameparts: new_frameparts,
+        }
+    }
+}
+
+#[derive(Clone, PrototypeElement)]
+pub struct Animation {
+    #[hidden_element]
+    pub frames: Vec<Frame>,
+}
+
+#[derive(Clone, PrototypeElement)]
+pub struct AnimationPair {
+    pub sprites: Arc<Sprite>,
+    pub actions: Arc<Actions>,
+}
+
+#[derive(Clone, PrototypeElement)]
+pub struct AnimationData {
+    pub animation_pair: Vec<AnimationPair>,
+    pub animations: Vec<Animation>,
+    pub delays: Vec<f32>,
+    #[hidden_element]
+    pub entity_type: EntityType,
+}
+
+// This function generates the convertion to the big rectangle

```

Comment:
This function looks like a screen space to NDC conversion! What is "big rectangle" in this context? Can you please explain this more?

Reply:
This is the conversion of "normalized" coordinates of a frame part inside the frame bounding box rectangle with vertex [-1, 0], [-1, 2], [1, 0], [1, 2]